multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/oauth2/KeycloakOAuthAuthenticationService.java

```diff
@@ -0,0 +1,149 @@
+   */
+  @GET
+  @Path("authenticate")
+  public Response authenticate(
+      @Required @QueryParam("oauth_provider") String oauthProvider,
+      @Required @QueryParam("redirect_after_login") String redirectAfterLogin,
+      @Context HttpServletRequest request)
+      throws ForbiddenException, BadRequestException {
+
+    Jwt jwtToken = (Jwt) request.getAttribute("token");
+    if (jwtToken == null) {
+      throw new BadRequestException("No token provided.");
+    }
+    DefaultClaims token = (DefaultClaims) jwtToken.getBody();
+    final String clientId = token.getAudience();
+    final String nonce = UUID.randomUUID().toString();
+    final String sessionState = token.get("session_state", String.class);
+    MessageDigest md;
+    try {
+      md = MessageDigest.getInstance("SHA-256");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+    final String input = nonce + sessionState + clientId + oauthProvider;
+    byte[] check = md.digest(input.getBytes(StandardCharsets.UTF_8));
+    final String hash = Base64Url.encode(check);
+    request.getSession().setAttribute("hash", hash); // TODO: for what?
+    String accountLinkUrl =
+        KeycloakUriBuilder.fromUri(keycloakConfiguration.get().get(AUTH_SERVER_URL_SETTING))
+            .path("/realms/{realm}/broker/{provider}/link")
+            .queryParam("nonce", nonce)
+            .queryParam("hash", hash)
+            .queryParam("client_id", clientId)
+            .queryParam("redirect_uri", redirectAfterLogin)
+            .build(keycloakConfiguration.get().get(REALM_SETTING), oauthProvider)
+            .toString();
+    return Response.temporaryRedirect(URI.create(accountLinkUrl)).build();
+  }
+
+  /**
+   * Gets OAuth token for user.
+   *
+   * @param oauthProvider OAuth provider name
+   * @return OAuthToken
+   * @throws ServerException
+   */
+  @GET
+  @Path("token")
+  @Produces(MediaType.APPLICATION_JSON)
+  public OAuthToken token(@Required @QueryParam("oauth_provider") String oauthProvider)
+      throws ForbiddenException, BadRequestException, ConflictException, NotFoundException,
+          ServerException, UnauthorizedException {
+
+    try {
+      String token =
+          requestFactory
+              .fromUrl(
+                  KeycloakUriBuilder.fromUri(
+                          keycloakConfiguration.get().get(AUTH_SERVER_URL_SETTING))
+                      .path("/realms/{realm}/broker/{provider}/token")
+                      .build(keycloakConfiguration.get().get(REALM_SETTING), oauthProvider)
+                      .toString())
+              .request()
+              .asString();
+      Map<String, String> params = splitQuery(token);
+      return DtoFactory.newDto(OAuthToken.class)
+          .withToken(params.get("access_token"))
+          .withScope(params.get("scope"));
+    } catch (IOException e) {
+      throw new ServerException(e.getMessage());
+    }
+  }
+
```

Comment:
UrlUtils.getQueryParameters

Reply:
It is not suitable here. It requires input to be valid URL (with scheme, hostname...) but we have only key=value pairs.